jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/json/JsonCfnParser.kt

```diff
@@ -0,0 +1,395 @@
+import com.intellij.json.psi.JsonReferenceExpression
+import com.intellij.json.psi.JsonStringLiteral
+import com.intellij.json.psi.JsonValue
+import com.intellij.psi.PsiElement
+import com.intellij.util.containers.BidirectionalMap
+import software.aws.toolkits.jetbrains.services.cloudformation.CfnArrayValueNode
+import software.aws.toolkits.jetbrains.services.cloudformation.CfnConditionNode
+import software.aws.toolkits.jetbrains.services.cloudformation.CfnConditionsNode
+import software.aws.toolkits.jetbrains.services.cloudformation.CfnExpressionNode
+import software.aws.toolkits.jetbrains.services.cloudformation.CfnFirstLevelMappingNode
+import software.aws.toolkits.jetbrains.services.cloudformation.CfnFunctionNode
+import software.aws.toolkits.jetbrains.services.cloudformation.CfnGlobalsNode
+import software.aws.toolkits.jetbrains.services.cloudformation.CfnIntrinsicFunction
+import software.aws.toolkits.jetbrains.services.cloudformation.CfnMappingValue
+import software.aws.toolkits.jetbrains.services.cloudformation.CfnMappingsNode
+import software.aws.toolkits.jetbrains.services.cloudformation.CfnMetadataNode
+import software.aws.toolkits.jetbrains.services.cloudformation.CfnNameValueNode
+import software.aws.toolkits.jetbrains.services.cloudformation.CfnNamedNode
+import software.aws.toolkits.jetbrains.services.cloudformation.CfnNode
+import software.aws.toolkits.jetbrains.services.cloudformation.CfnObjectValueNode
+import software.aws.toolkits.jetbrains.services.cloudformation.CfnOutputNode
+import software.aws.toolkits.jetbrains.services.cloudformation.CfnOutputsNode
+import software.aws.toolkits.jetbrains.services.cloudformation.CfnParameterNode
+import software.aws.toolkits.jetbrains.services.cloudformation.CfnParametersNode
+import software.aws.toolkits.jetbrains.services.cloudformation.CfnParsedFile
+import software.aws.toolkits.jetbrains.services.cloudformation.CfnProblem
+import software.aws.toolkits.jetbrains.services.cloudformation.CfnResourceConditionNode
+import software.aws.toolkits.jetbrains.services.cloudformation.CfnResourceDependsOnNode
+import software.aws.toolkits.jetbrains.services.cloudformation.CfnResourceNode
+import software.aws.toolkits.jetbrains.services.cloudformation.CfnResourcePropertiesNode
+import software.aws.toolkits.jetbrains.services.cloudformation.CfnResourcePropertyNode
+import software.aws.toolkits.jetbrains.services.cloudformation.CfnResourceTypeNode
+import software.aws.toolkits.jetbrains.services.cloudformation.CfnResourcesNode
+import software.aws.toolkits.jetbrains.services.cloudformation.CfnRootNode
+import software.aws.toolkits.jetbrains.services.cloudformation.CfnScalarValueNode
+import software.aws.toolkits.jetbrains.services.cloudformation.CfnSecondLevelMappingNode
+import software.aws.toolkits.jetbrains.services.cloudformation.CfnSection
+import software.aws.toolkits.jetbrains.services.cloudformation.CfnTransformNode
+
+class JsonCfnParser private constructor() {
```

Comment:
Are the JSON and YAML parsers based on any common interface? should they be? (although I guess the interface would probably be `parse(T)->CfnFile?` which isn't that useful)

Reply:
The original code didn't have one, but I am curious if I can take a page out of `JsonLikePsiWalker` and build an adapter to share most of the code.